crates/pallet-receipts/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+    #[pallet::storage]
+    pub type ReceiptVotes<T: Config> = StorageNMap<
+        _,
+        (
+            NMapKey<Twox64Concat, DomainId>,
+            NMapKey<Twox64Concat, T::Hash>,
+            NMapKey<Twox64Concat, H256>,
+        ),
+        u32,
+        ValueQuery,
+    >;
+
+    /// Mapping for tracking the secondary state roots.
+    ///
+    /// (core_domain_id, core_block_number, core_block_hash, core_state_root) -> state_root
+    #[pallet::storage]
+    #[pallet::getter(fn state_root)]
+    pub(super) type StateRoots<T: Config> = StorageNMap<
+        _,
+        (
+            NMapKey<Twox64Concat, DomainId>,
+            NMapKey<Twox64Concat, T::BlockNumber>,
+            NMapKey<Twox64Concat, T::DomainHash>,
+        ),
+        T::DomainHash,
+        OptionQuery,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A new domain receipt have processed
+        NewDomainReceipt {
+            domain_id: DomainId,
+            primary_number: T::BlockNumber,
+            primary_hash: T::Hash,
+        },
+        /// A fraud proof was processed.
+        FraudProofProcessed {
+            domain_id: DomainId,
+            new_best_number: T::BlockNumber,
+            new_best_hash: T::Hash,
+        },
+    }
+}
+
+pub enum Error {
+    /// The parent execution receipt is missing.
+    MissingParent,
+}
+
+impl<T: Config> Pallet<T> {
+    /// Returns the block number of the latest receipt.
+    pub fn head_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        let (_, best_number) = <ReceiptHead<T>>::get(domain_id);
+        best_number
+    }
+
+    /// Returns the block number of the oldest receipt still being tracked in the state.
+    pub fn oldest_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        Self::finalized_receipt_number(domain_id) + One::one()
+    }
+
+    /// Returns the block number of latest _finalized_ receipt.
+    pub fn finalized_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        let (_, best_number) = <ReceiptHead<T>>::get(domain_id);
+        best_number.saturating_sub(T::ReceiptsPruningDepth::get())
+    }
+
+    /// Initialize the genesis execution receipt
+    pub fn initialize_genesis_receipt(domain_id: DomainId, genesis_hash: T::Hash) {
+        let genesis_receipt = ExecutionReceipt {
+            primary_number: Zero::zero(),
+            primary_hash: genesis_hash,
+            domain_hash: T::DomainHash::default(),
+            trace: Vec::new(),
+            trace_root: Default::default(),
+        };
+        Self::apply_new_best_receipt(domain_id, &genesis_receipt);
+        // Explicitly initialize the oldest receipt number even not necessary as ValueQuery is used.
+        <OldestReceiptNumber<T>>::insert::<_, T::BlockNumber>(domain_id, Zero::zero());
+    }
+
+    /// Track the execution receipts for the domain
+    pub fn track_receipts(
```

Comment:
Can be easier to review if these pure function abstractions(`track_receipts`, `submit_fraud_proof`, `remove_expired_receipts`) can be included in the first refactoring commit.